tuple.go

```diff
@@ -0,0 +1,424 @@
+}
+
+func (t Tuple) Bytes() []byte {
+	panic("allows the Tuple type to conform to the Encodable interface")
+}
+
+func EncodeTuple(t interface{}, buffer *bytes.Buffer) {
+	tVal := reflect.ValueOf(t)
+
+	if tVal.Kind() != reflect.Struct {
+		panic("not a SCALE Tuple type")
+	}
+
+	// Tinygo does not support: reflect.VisibleFields(tVal.Type())
+	for i := 0; i < tVal.NumField(); i++ {
+		field := tVal.Field(i)
+
+		if reflect.TypeOf(t).Field(i).IsExported() {
+
+			switch field.Kind() {
+			case reflect.Bool:
+				ConvertTo[Bool](field).Encode(buffer)
+			case reflect.Uint, reflect.Int:
+				ConvertTo[Compact](field).Encode(buffer)
+			case reflect.Uint8:
+				ConvertTo[U8](field).Encode(buffer)
+			case reflect.Int8:
+				ConvertTo[I8](field).Encode(buffer)
+			case reflect.Uint16:
+				ConvertTo[U16](field).Encode(buffer)
+			case reflect.Int16:
+				ConvertTo[I16](field).Encode(buffer)
+			case reflect.Uint32:
+				ConvertTo[U32](field).Encode(buffer)
+			case reflect.Int32:
+				ConvertTo[I32](field).Encode(buffer)
+			case reflect.Uint64:
+				ConvertTo[U64](field).Encode(buffer)
+			case reflect.Int64:
+				ConvertTo[I64](field).Encode(buffer)
+			case reflect.String:
+				ConvertTo[Str](field).Encode(buffer)
+			case reflect.Array:
+				// U128, I128
+				switch field.Type() {
+				case reflect.TypeOf(*new(U128)):
+					ConvertTo[U128](field).Encode(buffer)
+				case reflect.TypeOf(*new(I128)):
+					ConvertTo[I128](field).Encode(buffer)
+				default:
+					panic("unreachable case (Array) in EncodeTuple")
+				}
+			case reflect.Slice:
+				// Sequence[T], FixedSequence[T], VaryingData
+				SequenceFieldEncode(field, buffer)
+			case reflect.Map:
+				DictionaryFieldEncode(field, buffer)
+			case reflect.Struct:
+				// Empty, CompactU128
+				switch field.Type() {
+				case reflect.TypeOf(*new(Empty)):
+					EncodeTuple(field.Interface(), buffer)
+				case reflect.TypeOf(*new(CompactU128)):
+					ConvertTo[CompactU128](field).Encode(buffer)
+				default:
+					// Option[T], Result[T], Tuple
+					if field.Kind() == reflect.Struct {
+						EncodeTuple(field.Interface(), buffer)
+					} else {
+						panic("unreachable case (Struct) in EncodeTuple")
+					}
+				}
+			case reflect.Float32:
+				panic("encoding of Float32 field is not supported")
+			case reflect.Float64:
+				panic("encoding of Float64 field is not supported")
+			case reflect.Complex64:
+				panic("encoding of Complex64 field is not supported")
+			case reflect.Complex128:
+				panic("encoding of Complex128 field is not supported")
+			case reflect.Uintptr:
+				panic("encoding of Uintptr field is not implemented")
+			case reflect.UnsafePointer:
+				panic("encoding of UnsafePointer field is not implemented")
+			case reflect.Pointer:
+				panic("encoding of Pointer field is not implemented")
+			case reflect.Chan:
+				panic("encoding of Chan field is not implemented")
+			case reflect.Func:
+				panic("encoding of Func field is not implemented")
```

Comment:
nit: we can match those in one case with `panic("not implemented")`

Reply:
👍